source/opt/scalar_analysis_nodes.h

```diff
@@ -0,0 +1,284 @@
+#include <string>
+#include <vector>
+#include "opt/tree_iterator.h"
+
+namespace spvtools {
+namespace ir {
+class Loop;
+}  // namespace ir
+
+namespace opt {
+
+class ScalarEvolutionAnalysis;
+class SEConstantNode;
+class SERecurrentNode;
+class SEAddNode;
+class SEMultiplyNode;
+class SENegative;
+class SEValueUnknown;
+class SECantCompute;
+
+// Abstract class representing a node in the scalar evolution DAG. Each node
+// contains a vector of pointers to its children and each subclass of SENode
+// implements GetType and an As method to allow casting. SENodes can be hashed
+// using the SENodeHash functor. The vector of children is sorted when a node is
+// added. This is important as it allows the hash of X+Y to be the same as Y+X.
+class SENode {
+ public:
+  enum SENodeType {
+    Constant,
+    RecurrentExpr,
+    Add,
+    Multiply,
+    Negative,
+    ValueUnknown,
+    CanNotCompute
+  };
+
+  explicit SENode(opt::ScalarEvolutionAnalysis* parent_analysis)
+      : parent_analysis_(parent_analysis) {}
+
+  virtual SENodeType GetType() const = 0;
+
+  virtual ~SENode() {}
+
+  inline void AddChild(SENode* child) {
+    children_.push_back(child);
+
+    // Children are sorted so the hashing
+    std::sort(children_.begin(), children_.end());
+  }
+
+  // Get the type as an std::string. This is used to represent the node in the
+  // dot output and is used to hash the type as well.
+  std::string AsString() const;
+
+  // Dump the SENode and its immediate children, if |recurse| is true then it
+  // will recurse through all children to print the DAG starting from this node
+  // as a root.
+  void DumpDot(std::ostream& out, bool recurse = false) const;
+
+  // Checks if two nodes are the same by hashing them.
+  bool operator==(const SENode& other) const;
+
+  // Checks if two nodes are not the same by comparing the hashes.
+  bool operator!=(const SENode& other) const;
+
+  // Return the child node at |index|.
+  inline SENode* GetChild(size_t index) { return children_[index]; }
+  inline const SENode* GetChild(size_t index) const { return children_[index]; }
+
+  // Iterator to iterate over the child nodes.
+  using iterator = std::vector<SENode*>::iterator;
+  using const_iterator = std::vector<SENode*>::const_iterator;
+
+  // Iterate over immediate child nodes.
+  iterator begin() { return children_.begin(); }
+  iterator end() { return children_.end(); }
+
+  // Constant overloads for iterating over immediate child nodes.
+  const_iterator begin() const { return children_.cbegin(); }
+  const_iterator end() const { return children_.cend(); }
+  const_iterator cbegin() { return children_.cbegin(); }
+  const_iterator cend() { return children_.cend(); }
+
+  // Iterator to iterate over the entire DAG. Even though we are using the tree
+  // iterator it should still be safe to iterate over. However, nodes with
+  // multiple parents will be visited multiple times, unlike in a tree.
+  using dag_iterator = TreeDFIterator<SENode>;
```

Comment:
Is there a reason you want to traverse the same node multiple times?  This leads to traversing every path in the DAG, which is exponential in the worst case.  If you don't need it, we should not do it.

Reply:
Yes, when traversing a node we want to visit all its children. If it has the same node as a child twice then we want to visit that node twice